# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [3]:
len(data['days_employed'])

21525

In [4]:
data1 = data[(data['days_employed'].isna())].head(30)

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [6]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [7]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [8]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [9]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [10]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [11]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [12]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [13]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [14]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [15]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [16]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [17]:
data.duplicated().sum()

71

In [18]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [19]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [22]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Изначально сгруппируем данные по долгу между людьми с разным количеством детей

In [24]:
data.groupby('children')['debt'].sum()



children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

Цифры выше нерепрезентативны, поскольку не учитывают число людей с разным количеством детей. Найдем число людей для разного количества детей


In [25]:
data.groupby('children')['children'].count()


children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

Зависимость оценим, посчитав конверсию


In [26]:
data.groupby('children')['debt'].sum()/data.groupby('children')['children'].count()


children
0    0.075438
1    0.092346
2    0.094542
3    0.081818
4    0.097561
5    0.000000
dtype: float64

**Вывод:** 
Зависимость есть.
Хуже всего возвращают кредиты в срок люди с 4 детьми.  Наименьшая конверсия у людей с 5 детьми, что скорее всего обусловлено их малым количестве в таблице (т.е. недостаточной статистикой). Потому что даже если один человек из этой категории не вернет кредит вовремя, то конверсия составит 1/10 = 0.1 и конверсия будет максимальной из представленных. Поэтому, если не учитывать эту категорию, получим, что минимальная конверсия соответствует людям без детей.
Также наблюдаем локальный минимум для семей с 3 детьми - это обусловлено возможно появлением льгот для многодетных семей (от 3 детей) и финансовая нагрузка не такая большая. 
В то же время люди с 4 детьми имеют высокую конверсию - можно предположить, что несмотря на действующие в стране льготы для семей с 3 детьми и выше, 4 ребенок добавляет существенную финансовую нагрузку на семью.

**Таким образом:** 
1) если не учитывать категорию с 5 детьми можно сделать вывод, что само по себе наличие детей влияет на возврат кредита в срок - люди чаще отдают кредиты вовремя.

2) неплохую конверсию показыют люди с 3 детьми (минимальным числом детей для статуса многодетной семьи), поскольку возможно финансовая нагрузка облегчается существующими льготами.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Проводим аналогичные расчеты, но с использованием столбца 'family_status'

In [27]:
data.groupby('family_status')['family_status'].count()

family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4134
женат / замужем          12261
Name: family_status, dtype: int64

In [28]:
data.groupby('family_status')['debt'].sum()

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64

In [29]:
data.groupby('family_status')['debt'].sum()/data.groupby('family_status')['family_status'].count()

family_status
Не женат / не замужем    0.097639
в разводе                0.070648
вдовец / вдова           0.066246
гражданский брак         0.093130
женат / замужем          0.075606
dtype: float64

**Вывод:** 
Зависимость есть. 

Наименьшая конверсия характерна для категории "вдовец/вдова". Чаще всего не отдают кредиты в срок неженатые и незамужние.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Проводим аналогичные расчеты, но с использованием столбца 'total_income_category'

In [30]:
data.groupby('total_income_category')['total_income_category'].count()

total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: total_income_category, dtype: int64

In [31]:
data.groupby('total_income_category')['debt'].sum()

total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: debt, dtype: int64

In [32]:
data.groupby('total_income_category')['debt'].sum() / \
data.groupby('total_income_category')['total_income_category'].count()

total_income_category
A    0.080000
B    0.070602
C    0.084982
D    0.060172
E    0.090909
dtype: float64

**Вывод:** 
Зависимость есть, но необратнопропорциональная доходу (т.е. конверсия не снижается при увеличении дохода).
Наихудшая (наибольшая) конверсия характерная для категории 'Е', т.е. люди с доходом от 0 до 30000 чаще всего не отдают кредит в срок. Наименьшая конверсия соответствует категории 'D'. В то же время статистических данных для категорий 'D','Е' и 'А' мало по сравнению с остальными категориями, что может стать причиной неправильной оценки. 

In [33]:
data2 = data.groupby('total_income_category').agg({'debt':['sum','mean'], 'total_income_category': 'count'})
data2

debt           total_income_category
                        sum      mean                 count
total_income_category                                      
A                         2  0.080000                    25
B                       354  0.070602                  5014
C                      1353  0.084982                 15921
D                        21  0.060172                   349
E                         2  0.090909                    22

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Проводим аналогичные расчеты, но с использованием столбца 'purpose_category'

In [34]:
data.groupby('purpose_category')['purpose_category'].count()

purpose_category
операции с автомобилем       4279
операции с недвижимостью    10751
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

In [35]:
data.groupby('purpose_category')['debt'].sum()

purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

In [36]:
data.groupby('purpose_category')['debt'].sum()/data.groupby('purpose_category')['purpose_category'].count()

purpose_category
операции с автомобилем      0.093480
операции с недвижимостью    0.072551
получение образования       0.092528
проведение свадьбы          0.079118
dtype: float64

**Вывод:** 
Чаще всего не отдают кредит в срок для двух целей: получение образования и операции с автомобилем. Кредит на недвижимость погашают в срок чаще по сравнению с другими целями.

In [37]:
data_pivot = data.pivot_table(index=['purpose_category'], columns='family_status', values='debt', aggfunc='mean')
data_pivot

family_status             Не женат / не замужем  в разводе  вдовец / вдова  \
purpose_category                                                             
операции с автомобилем                 0.129134   0.075269        0.091743   
операции с недвижимостью               0.081337   0.068452        0.052336   
получение образования                  0.107826   0.071429        0.075758   
проведение свадьбы                          NaN        NaN             NaN   

family_status             гражданский брак  женат / замужем  
purpose_category                                             
операции с автомобилем            0.118605         0.083180  
операции с недвижимостью          0.092105         0.069727  
получение образования             0.148883         0.083528  
проведение свадьбы                0.079118              NaN

Из полученной сводной таблицы видно, что доля невозвратов кредитов в срок различна для каждой из целей и также зависит от семейного положения.

Вот еще пример расчета конверсии невозвратов с учетом дополнительных зависимостей - конверсия по возврату кредита в срок от семейного положения с учетом числа детей у людей

In [38]:
#data.groupby(['children', 'family_status'])['debt'].sum()/ \
#data.groupby(['children', 'family_status'])['debt'].count()

In [39]:
data_pivot = data.pivot_table(index=['children'], columns='family_status', values='debt', aggfunc='mean')
data_pivot

family_status  Не женат / не замужем  в разводе  вдовец / вдова  \
children                                                          
0                           0.092838   0.070153        0.062574   
1                           0.115813   0.067308        0.090909   
2                           0.120000   0.086420        0.150000   
3                           0.125000   0.090909        0.000000   
4                           0.500000   0.000000        0.000000   
5                                NaN        NaN             NaN   

family_status  гражданский брак  женат / замужем  
children                                          
0                      0.083883         0.069095  
1                      0.118593         0.082689  
2                      0.087464         0.094586  
3                      0.142857         0.068273  
4                      0.000000         0.103448  
5                      0.000000         0.000000

 Для ситуации с отсутствием детей наименьшая конверсия также характерна для семейного положения вдовец/вдова.
 
 Для ситуации с 1 ребенком наименьшая конверсия характерна для семейного положения в разводе.
 
 Для 2 детей - гражд брак и в разводе и т.д.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Изначально пропуски данных были в двух столбцах: 'days_employed' и 'total_income' по 2174 в каждом (т.е. минимум в каждом 10 значении: 21525/2174, что много для человеческого фактора). Причем значение NaN по столбцам встречались в одних и тех же строках. Значения в остальных строках не коррелируют с пропусками:

In [40]:
data1

children  days_employed  dob_years            education  education_id  \
12          0            NaN         65              среднее             1   
26          0            NaN         41              среднее             1   
29          0            NaN         63              среднее             1   
41          0            NaN         50              среднее             1   
55          0            NaN         54              среднее             1   
65          0            NaN         21              среднее             1   
67          0            NaN         52               высшее             0   
72          1            NaN         32               высшее             0   
82          2            NaN         50               высшее             0   
83          0            NaN         52              среднее             1   
90          2            NaN         35               высшее             0   
94          1            NaN         34               высшее             0   
96          0            NaN         44              СРЕДНЕЕ             1   
97          0            NaN         47               высшее             0   
120         0            NaN         46              среднее             1   
121         0            NaN         29               высшее             0   
135         0            NaN         27              среднее             1   
141         0            NaN         39              среднее             1   
145         0            NaN         62              среднее             1   
174         0            NaN         55               высшее             0   
181         0            NaN         26              среднее             1   
189         1            NaN         30              среднее             1   
205         1            NaN         31               высшее             0   
220         1            NaN         23  неоконченное высшее             2   
241         0            NaN         47              среднее             1   
242         0            NaN         58              среднее             1   
247         1            NaN         60               высшее             0   
250         0            NaN         54               высшее             0   
264         2            NaN         40              среднее             1   
278         1            NaN         23              Среднее             1   

             family_status  family_status_id gender  income_type  debt  \
12        гражданский брак                 1      M    пенсионер     0   
26         женат / замужем                 0      M  госслужащий     0   
29   Не женат / не замужем                 4      F    пенсионер     0   
41         женат / замужем                 0      F  госслужащий     0   
55        гражданский брак                 1      F    пенсионер     1   
65   Не женат / не замужем                 4      M    компаньон     0   
67         женат / замужем                 0      F    пенсионер     0   
72         женат / замужем                 0      M  госслужащий     0   
82         женат / замужем                 0      F    сотрудник     0   
83         женат / замужем                 0      M    сотрудник     0   
90         женат / замужем                 0      F    сотрудник     0   
94        гражданский брак                 1      F    компаньон     0   
96         женат / замужем                 0      F    сотрудник     0   
97         женат / замужем                 0      F    сотрудник     0   
120        женат / замужем                 0      F    сотрудник     0   
121        женат / замужем                 0      F    компаньон     0   
135        женат / замужем                 0      M    компаньон     0   
141       гражданский брак                 1      M    сотрудник     0   
145        женат / замужем                 0      M    пенсионер     0   
174         вдовец / вдова                 2      F    компаньон     0   
181       гражданский брак      

**Вывод:** Скорее всего имел место технологический фактор - сбой в процессе копирования, считывания, смены формата или записи данных.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
На примере решаемой задачи мы столкнулись с ситуацией, когда в столбце 'days_employed' было 2 аномально больших значения. Если бы мы заполнили пропуски в этом столбце не медианными, а средними значениями по каждому типу занятости 'income_type', мы бы получили смещенное в большую сторону значение, которое бы хуже отражало реальные данные.

### Шаг 4: общий вывод.

Изначально в описании проекта была задача - разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

В ходе решения задачи получены конверсии по возврату кредитов для людей с разных количеством детей, разным семейным положением, разным доходом и для разных целей. Во всех случаях наблюдалась определенная зависимость: возврат кредита зависит от всех перечисленных факторов. 

В ряде случаев значение конверсии нельзя назвать объективным из-за малого объема статистики по случаям, что может быть причиной неправильных выводов. 

Кроме того, в проекте были исследованы простые зависимости. Если мы зададимся дополнительными зависимостями, например, посчитаем конверсию по возврату кредита в срок от семейного положения с учетом числа детей у людей, то ситуация изменится (см. расчет выше, где применял метод pivot_table())
